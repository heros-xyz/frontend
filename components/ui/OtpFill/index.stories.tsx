import type { Meta, StoryObj } from "@storybook/react";
import OtpFill from ".";

const meta: Meta<typeof OtpFill> = {
  title: "Components/OtpFill",
  component: OtpFill,
  tags: ["autodocs"],
  parameters: {
    layout: "fullscreen",
  },
};

export default meta;
type Story = StoryObj<typeof OtpFill>;

export const OtpFillComponent: Story = {
  args: {
    title: "OTP Verification",
    description: "We've emailed an OTP to you. Please enter the code here.",
    diffCount: 5,
    textButton: "VERIFY",
  },
};
