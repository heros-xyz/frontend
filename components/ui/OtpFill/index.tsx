import {
  Box,
  Button,
  HStack,
  Link,
  PinInput,
  PinInputField,
  Text,
} from "@chakra-ui/react";
import { useEffect, useState } from "react";
import { useFormik } from "formik";
import * as yup from "yup";
import ErrorMessage from "@/components/common/ErrorMessage";
import { getAuthErrorCode } from "@/utils/constants";
import { isEmptyObject } from "@/utils/functions";
import useCountdown from "../../../hooks/useCountdown";

interface OtpProps {
  title: string;
  description: string;
  validTime: number;
  textButton: string;
  errorMessage?: string | number;
  isLoading?: boolean;
  otpValue: string;
  onSubmit: (otp: string) => void;
  resendOtp?: () => void;
}

const OtpFill: React.FC<OtpProps> = ({
  title,
  description,
  validTime,
  textButton,
  errorMessage,
  isLoading,
  otpValue,
  onSubmit,
  resendOtp,
}) => {
  const [verifyOtpError, setVerifyOtpError] = useState<
    string | undefined | number
  >("");
  const { minutes, seconds, reset } = useCountdown(validTime);

  const formik = useFormik({
    initialValues: {
      otp: otpValue,
    },
    validationSchema: yup.object().shape({
      otp: yup
        .string()
        .test(
          "otp-length",
          "OTP must contain 6 characters, please recheck!",
          (val) => {
            return val?.length === 6;
          }
        ),
    }),
    onSubmit: (values) => {
      onSubmit(values.otp);
    },
  });

  const onResend = () => {
    reset();
    resendOtp && resendOtp();
  };

  useEffect(() => {
    formik.setFieldValue("otp", otpValue);
  }, [otpValue]);

  // useEffect(() => {
  //   isEmptyObject(formik.errors) && formik.handleSubmit();
  // }, [formik.errors]);

  useEffect(() => {
    setVerifyOtpError(errorMessage);
  }, [errorMessage]);

  return (
    <Box
      bg="primary"
      minH="100vh"
      display="flex"
      justifyContent="center"
      color="white"
    >
      <form onSubmit={formik.handleSubmit}>
        <Box
          w={"full"}
          display="flex"
          flexDirection="column"
          maxWidth="500px"
          pt={{ base: "70px", xl: "120px" }}
          paddingX={{ base: "20px", xl: "0px" }}
        >
          <Box textAlign="center" mb={{ base: "40px", xl: "50px" }}>
            <Box
              fontWeight="extrabold"
              fontSize={{ base: "2rem", xl: "5xl" }}
              textTransform="capitalize"
              lineHeight="140%"
              mb={{ base: "10px", xl: "15px" }}
            >
              {title}
            </Box>
            <Box
              fontWeight="normal"
              fontSize={{ base: "md", xl: "xl" }}
              lineHeight="140%"
            >
              {description}
            </Box>
          </Box>

          <HStack
            mb={{ base: 2, xl: 3 }}
            w="full"
            display="grid"
            gridTemplateColumns="repeat(6, 1fr)"
            justifyContent="space-between"
            gap={{ base: "15px", xl: "20px" }}
          >
            <PinInput
              size="lg"
              variant="flushed"
              placeholder=""
              value={formik.values.otp}
              onChange={(value) => {
                setVerifyOtpError("");
                formik.setFieldValue("otp", value);
              }}
              autoFocus
            >
              <PinInputField
                fontSize="32px"
                pb="15px"
                w="full"
                ml="0px !important"
                autoFocus
              />
              <PinInputField
                fontSize="32px"
                pb="15px"
                w="full"
                ml="0px !important"
              />
              <PinInputField
                fontSize="32px"
                pb="15px"
                w="full"
                ml="0px !important"
              />
              <PinInputField
                fontSize="32px"
                pb="15px"
                w="full"
                ml="0px !important"
              />
              <PinInputField
                fontSize="32px"
                pb="15px"
                w="full"
                ml="0px !important"
              />
              <PinInputField
                fontSize="32px"
                pb="15px"
                w="full"
                ml="0px !important"
              />
            </PinInput>
          </HStack>

          <Box textAlign="left">
            <ErrorMessage
              condition={!!verifyOtpError}
              errorMessage={getAuthErrorCode(verifyOtpError)}
            />
          </Box>
          <Box
            mb="50px"
            textAlign="center"
            fontSize={{ base: "md", xl: "xl" }}
            mt={{ base: "40px", xl: "50px" }}
          >
            <Box
              display="flex"
              justifyContent="center"
              fontWeight="normal"
              lineHeight="140%"
              mb={3}
            >
              <Text>Valid for: &nbsp;</Text>
              <Box color="secondary" lineHeight="140%">
                <Text> {`${minutes}:${seconds}`}</Text>
              </Box>
            </Box>
            <Link
              color="accent.1"
              textDecorationLine="underline"
              fontWeight="medium"
              onClick={onResend}
            >
              Resend OTP
            </Link>
          </Box>
          <Box
            w={{ base: "full", lg: "100%" }}
            display="flex"
            justifyContent="center"
          >
            <Button
              fontSize={{ base: "md", xl: "xl" }}
              variant="secondary"
              fontWeight="bold"
              lineHeight="140%"
              w={{ base: "full", lg: "fit-content" }}
              opacity={isEmptyObject(formik.errors) ? 1 : 0.5}
              type="submit"
              isLoading={isLoading}
              // cursor="not-allowed"
            >
              {textButton}
            </Button>
          </Box>
        </Box>
      </form>
    </Box>
  );
};

export default OtpFill;
