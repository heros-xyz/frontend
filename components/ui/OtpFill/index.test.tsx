import { render } from "@testing-library/react";
import OtpFill from ".";

it("renders homepage unchanged", () => {
  const { container } = render(
    <OtpFill
      title="OTP Verification"
      description="We've emailed an OTP to you. Please enter the code here."
      textButton="VERIFY"
      validTime={5}
    />
  );
  expect(container).toMatchSnapshot();
});
