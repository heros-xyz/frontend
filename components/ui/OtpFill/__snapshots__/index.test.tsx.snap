// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <div
    class="css-1auwmo4"
  >
    <form>
      <div
        class="css-1v18uo6"
      >
        <div
          class="css-3o270"
        >
          <div
            class="css-1job21j"
          >
            OTP Verification
          </div>
          <div
            class="css-1vrpgvd"
          >
            We've emailed an OTP to you. Please enter the code here.
          </div>
        </div>
        <div
          class="chakra-stack css-2ds2n6"
        >
          <input
            aria-label="Please enter your pin code"
            autocomplete="off"
            class="chakra-pin-input css-61fpcf"
            data-index="0"
            id="pin-input-:r0:-0"
            inputmode="numeric"
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autocomplete="off"
            class="chakra-pin-input css-61fpcf"
            data-index="1"
            id="pin-input-:r0:-1"
            inputmode="numeric"
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autocomplete="off"
            class="chakra-pin-input css-61fpcf"
            data-index="2"
            id="pin-input-:r0:-2"
            inputmode="numeric"
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autocomplete="off"
            class="chakra-pin-input css-61fpcf"
            data-index="3"
            id="pin-input-:r0:-3"
            inputmode="numeric"
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autocomplete="off"
            class="chakra-pin-input css-61fpcf"
            data-index="4"
            id="pin-input-:r0:-4"
            inputmode="numeric"
            placeholder=""
            type="tel"
            value=""
          />
          <input
            aria-label="Please enter your pin code"
            autocomplete="off"
            class="chakra-pin-input css-61fpcf"
            data-index="5"
            id="pin-input-:r0:-5"
            inputmode="numeric"
            placeholder=""
            type="tel"
            value=""
          />
        </div>
        <div
          class="css-13brihr"
        />
        <div
          class="css-k5a7t"
        >
          <div
            class="css-2noc44"
          >
            <p
              class="chakra-text css-0"
            >
              Valid for: Â 
            </p>
            <div
              class="css-ou80uo"
            >
              <p
                class="chakra-text css-0"
              >
                 
                05:00
              </p>
            </div>
          </div>
          <a
            class="chakra-link css-ptpdh5"
          >
            Resend OTP
          </a>
        </div>
        <div
          class="css-16yehnp"
        >
          <button
            class="chakra-button css-1xa3frc"
            type="submit"
          >
            VERIFY
          </button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
